### PR TITLE
trpc 10 migration

### LINKED ISSUE #4: migrate to trpc v10
when will v10 be released?

### PR DESCRIPTION
**Pull request by dai-shi**
close #4

### CODE REVIEW

**Comment by dai-shi** [comment]
```js
  type TQueries = TRouter['_def']['queries'];
```
seems deprecated??

**Comment by sachinraja** [comment]
Since we're using a proxy now there's no need for us to filter for `queries` on the type level. We just map over the procedure record (object of procedures) and conditionally check if it's a query: https://github.com/trpc/trpc/blob/next/packages/react/src/createTRPCReact.tsx#L101-L113

**Comment by dai-shi** [comment]
Oh, so the "record" is now a recursive structure? This requires a TS skill. No utils are exported from vanilla client, right? Let me try then.

**Comment by sachinraja** [comment]
> Oh, so the "record" is now a recursive structure? This requires a TS skill. 

Yes. We export a `createProxy` util but that's for runtime logic. Types will have to be implemented in the recursive manner shown. `@trpc/react` is a good reference to use for those things

**Comment by dai-shi** [comment]
note that jotai-trpc doesn't use proxies, because they don't match well.
spent a few hours to come up with 802ae7f.
however, there remains a TS error. There must be something wrong.
this feels like v10 is really tough for non-proxy apis.

**Comment by dai-shi** [comment]
Hm, maybe we should make `path` to be `string[]` type.

**Comment by sachinraja** [comment]
> note that jotai-trpc doesn't use proxies, because they don't match well.

Is there a particular reason for this? I can implement the TypeScript part if necessary.

**Comment by dai-shi** [comment]
We shouldn't (can't) create atoms dynamically (because their stable references are important for identification). They should be defined outside components in advance. What's your suggestion under this assumption?

**Comment by sachinraja** [comment]
The atom wouldn't be created dynamically, it'd still be created outside components.
```ts
const pokemonAtom = atomWithQuery('pokemon.byId', (get) => [get(nameAtom)]);

const pokemonAtom = trpc.pokemon.byId.atomWithQuery(/* ... */)
```
Is there a difference between these two? The proxy one essentially just calls the first one under the hood.

If it's truly not possible or not a good idea I suggest this:
```ts
const pokemonAtom = atomWithQuery(trpc.pokemon.byId, (get) => [get(nameAtom)]);
```
So the proxy is now just an argument to the function. The benefit is that you get jump to definition to the backend.

**Comment by dai-shi** [comment]
```js
const pokemonAtom = trpc.pokemon.byId.atomWithQuery(/* ... */)
```
Yeah, that is possible. One question is how we name `trpc` and `atomWithQuery` there.

btw, what's the benefit over `atomWithQuery('pokemon.byId', ...)`?

**Comment by dai-shi** [comment]
(fwiw, one reason I started jotai is to not to use proxies..)

**Comment by sachinraja** [comment]
> btw, what's the benefit over atomWithQuery('pokemon.byId', ...)?

Jump to definition (backend function) and refactoring. You can rename the backend endpoint and it'll also rename itself on the client automatically using TypeScript's refactoring capabilities.

> (fwiw, one reason I started jotai is to not to use proxies..)

That makes sense but using proxies like this is different than using proxies like valtio. If proxies are a deal breaker, then the string literal method is fine.

**Comment by dai-shi** [comment]
> You can rename the backend endpoint and it'll also rename itself on the client automatically using TypeScript's refactoring capabilities.

That's interesting!

> using proxies like this is different than using proxies like valtio.

I know. And, to me, in JS, it's less necessary to use proxies. 

> If proxies are a deal breaker, then the string literal method is fine.

What I can do is to provide both and let users choose. I'd probably use defineProperty instead of proxy anyway.

**Comment by sachinraja** [comment]
> I'd probably use defineProperty instead of proxy anyway.

that's not possible since clients have no knowledge of the router at runtime. That info only exists on the type level.

**Comment by dai-shi** [comment]
I see. That's convincing me for string based api. I'll try proxy based api and provide both. Feels like proxy based api can give better TS experience.

**Comment by dai-shi** [comment]
> provide both.

I'll work on proxy-base api only, and see how it goes.

**Comment by dai-shi** [comment]
@sachinraja Is `inferObservableValue` exported from somewhere? Or, any workaround?

**Comment by sachinraja** [comment]
It's exported from `@trpc/server/observable`

**Comment by dai-shi** [comment]
@sachinraja Thanks.

Now, do you have `trpc-pokemon-v10` or something for v10?

**Comment by sachinraja** [comment]
Not yet, but I'll update it sometime this week.

**Comment by sachinraja** [comment]
Finally got around to doing it: https://github.com/sachinraja/trpc-pokemon

File: examples/02_authtoken/src/App.tsx [hunk 1]
```diff
@@ -1,29 +1,39 @@
...
+    ],
   }),
 );
 
-const pokemonAtom = atomWithQuery('pokemon.all', [], undefined, (get) =>
-  get(clientAtom),
+const pokemonAtom = trpc.pokemon.all.atomWithQuery(
+  undefined,
+  undefined,
+  (get) => get(clientAtom).pokemon.all as any, // FIXME avoid any
```
Comment (dai-shi) [inline examples/02_authtoken/src/App.tsx:36]: @sachinraja One remaining issue is this `FIXME`. Can you help??

Comment (sachinraja) [reply examples/02_authtoken/src/App.tsx:36]: I'm confused, what's the purpose of `getProcedure` when we already have the path (`pokemon.all`)?

Comment (dai-shi) [reply examples/02_authtoken/src/App.tsx:36]: We don't directly use React Context, so there should be an optional way to inject a custom client.

Check out v9 based API:
https://github.com/jotai-labs/jotai-trpc/blob/f56509e971db886c8c12f906e00ba792deef7879/examples/02_authtoken/src/App.tsx#L10-L27

Comment (dai-shi) [reply examples/02_authtoken/src/App.tsx:36]: On second thought, it's not nice in DX. Let me see if I can pass paths around.

**Comment by dai-shi** [comment]
@sachinraja Okay, all fixed. I think the remaining `any` in the library code is unavoidable.
Thanks for your help!

**Review by sachinraja — APPROVED** [review]
awesome!

**Pull request merged**